File: stage2/problem12.java
import java.util.LinkedList;
import java.util.List;

public class problem12 {
    
    private static List<Integer> primes;
    
    private static int divisibleTriangleNumber(int n) {
        int triangleNr = 1;
        int tr = 1;
        while (findFactors(tr) <= n) {
            System.out.println(tr+ " "+findFactors(tr));
            triangleNr++;
            tr = (triangleNr)*(triangleNr+1)/2;
        }
        return tr;
    }
    
    private static int findFactors(int number) {
        int pointer = 0;
        int factors = 1;
        int currentExp = 0;
        while (number > 1) {
            currentExp = 0;
            while (number%primes.get(pointer) == 0) {
                number /= primes.get(pointer);
                currentExp++;
            }
            if (isPrime(number)) {
                primes.add(number);
            }
            factors *= (currentExp+1);
            pointer++; // no need to search for the next prime
        } 
        return factors; 
    }
    
    private static Boolean isPrime(int x) {
        if (x == 1) {
            return false;
        }
        int i = 0; 
        while (i < primes.size()) {
            if (x%primes.get(i) == 0) {
                return false;
            }
            i++;
        }
        return true;
    }
    
    public static void main(String[] args) {
        primes = new LinkedList<Integer>();
        primes.add(2);
        System.out.println(divisibleTriangleNumber(500));
    }
}

File: stage2/problem12.js
const primes = [2];

function divisibleTriangleNumber(n) {
  var triangleNr = 1 //Math.ceil(n/3);
  var flag = false;
  var tr = 1;
  while (findFactors(tr) <= n) {
    triangleNr++;
    flag = !flag;
    if (flag) {
      tr = (triangleNr+1)/2;
      tr *= triangleNr;
    } else {
      tr = (triangleNr)/2;
      tr *= (triangleNr+1);
    }
  }
  return tr;
}

function findFactors(number) {
  var pointer = 0;
  var factors = 1;
  var currentExp = 0;
  while (number > 1) {
    currentExp = 0;
    while (number%primes[pointer] == 0) {
      number /= primes[pointer];const primes = [2];

function divisibleTriangleNumber(n) {
  var triangleNr = 1 //Math.ceil(n/3);
  var tr = 1;
  while (findFactors(tr) <= n) {
    triangleNr++;
    tr = (triangleNr)*(triangleNr+1)/2;
  }
  return tr;
}

function findFactors(number) {
  var pointer = 0;
  var factors = 1;
  var currentExp = 0;
  while (number > 1) {
    currentExp = 0;
    while (number%primes[pointer] == 0) {
      number /= primes[pointer];
      currentExp++;
    }
    if (isPrime(number)) {
      primes.push(number);
      //console.log("Primes: ", primes);
    }
    factors *= (currentExp+1);
    pointer++; 
  } 
  return factors; 
}

function isPrime(x) {
  if (x == 1) {
    return false;
  }
  var i = 0;
  while (i < primes.length) {
    if (x%primes[i] == 0) {
      return false;
    }
    i++;
  }
  return true;
}

//test times out, the java program works. Just takes an hour or so.
console.log(divisibleTriangleNumber(500))

      currentExp++;
    }
    if (isPrime(number)) {
      primes.push(number);
      //console.log("Primes: ", primes);
    }
    factors *= (currentExp+1);
    pointer++; 
  } 
  return factors; 
}

function isPrime(x) {
  if (x == 1) {
    return false;
  }
  var i = 0;
  while (i < primes.length) {
    if (x%primes[i] == 0) {
      return false;
    }
    i++;
  }
  return true;
}

console.log(divisibleTriangleNumber(500))